mithril-aggregator/src/artifact_builder/cardano_database.rs

```diff
@@ -0,0 +1,213 @@
+use std::path::{Path, PathBuf};
+
+use anyhow::{anyhow, Context};
+use async_trait::async_trait;
+use semver::Version;
+
+use mithril_common::{
+    entities::{
+        ArtifactsLocations, CardanoDatabase, CardanoDbBeacon, Certificate, CompressionAlgorithm,
+        ProtocolMessagePartKey, SignedEntityType,
+    },
+    StdResult,
+};
+
+use crate::artifact_builder::ArtifactBuilder;
+
+pub struct CardanoDatabaseArtifactBuilder {
+    db_directory: PathBuf, // TODO: temporary, will be accessed through another dependency instead of direct path.
+    cardano_node_version: Version,
+    compression_algorithm: CompressionAlgorithm,
+}
+
+impl CardanoDatabaseArtifactBuilder {
+    pub fn new(
+        db_directory: PathBuf,
+        cardano_node_version: &Version,
+        compression_algorithm: CompressionAlgorithm,
+    ) -> Self {
+        Self {
+            db_directory,
+            cardano_node_version: cardano_node_version.clone(),
+            compression_algorithm,
+        }
+    }
+}
+
+#[async_trait]
+impl ArtifactBuilder<CardanoDbBeacon, CardanoDatabase> for CardanoDatabaseArtifactBuilder {
+    async fn compute_artifact(
+        &self,
+        beacon: CardanoDbBeacon,
+        certificate: &Certificate,
+    ) -> StdResult<CardanoDatabase> {
+        let merkle_root = certificate
+            .protocol_message
+            .get_message_part(&ProtocolMessagePartKey::CardanoDatabaseMerkleRoot)
+            .ok_or(anyhow!(
+                "Can not find CardanoDatabaseMerkleRoot protocol message part in certificate"
+            ))
+            .with_context(|| {
+                format!(
+                    "Can not compute CardanoDatabase artifact for signed_entity: {:?}",
+                    SignedEntityType::CardanoDatabase(beacon.clone())
+                )
+            })?;
+        let total_db_size_uncompressed = compute_uncompressed_database_size(&self.db_directory)?;
+
+        let cardano_database = CardanoDatabase::new(
+            merkle_root.to_string(),
+            beacon,
+            total_db_size_uncompressed,
+            ArtifactsLocations::default(), // TODO: temporary default locations, will be injected in next PR.
+            self.compression_algorithm,
+            &self.cardano_node_version,
+        );
+
+        Ok(cardano_database)
+    }
+}
+
+/// Return the sum of the files size contained in the subdirectories 'immutable', 'ledger' and 'volatile'.
+fn compute_uncompressed_database_size(db_directory: &Path) -> StdResult<u64> {
+    let subdirs = ["immutable", "ledger", "volatile"];
```

Comment:
This list should be computed automatically from the path.
Also, you could have used a single recursive function to do that operation.

Reply:
I completely refactored the function based on the different comments on the PR: https://github.com/input-output-hk/mithril/pull/2159/files#diff-aab140d744a307dd040e3dba02956b21858cb54510ea96a58ab094124897da8eR71-R94